docs: initial post on llms + data (#7217)
ibis-project/ibis: docs/posts/llms-and-data/index.qmd

---
title: LLMs and data
author: "Cody Peterson"
date: "2023-09-27"
freeze: auto
categories:
    - blog
    - ai
    - llm
---

## Introduction

> LLMs have been around for a while, but the recent versions -- along with other innovations around their usage -- have given instruction-following, conversational bots that can perform tasks on behalf of the user. The thought of using natural language to transform and analyze data is obvious and attractive.

Reviewer: Can you wrap these lines using `gq` (vim) or `gw` (newish versions of neovim)

## Approaches

When discussed internally at Voltron Data and identified three primary approaches to applying LLMs to data today:

1. AI writes SQL
2. AI writes a UDF (Python or another language)
3. Use AI in a UDF

### Approach 1: AI writes SQL

> SoTA LLMs are decent at generating SQL on their own. We can be clever to handle errors, retries, etc. but in its simplest form we ask a LLM to generate SQL and call it. Let's start with a simple example:

Reviewer: Assuming `SoTA` means `State of the art`, let's expand the acronym.

```{python}
#| code-fold: true
import ibis
import marvin

from ibis.expr.schema import Schema
from ibis.expr.types.relations import Table

> from dotenv import load_dotenv

Reviewer: Maybe an annotation saying why this is necessary?

---
title: LLMs and data
author: "Cody Peterson"
date: "2023-09-26"
freeze: auto
categories:
    - blog
    - ai
    - llm
---

## Introduction

While large-language models (LLMs) have been around for years, recent versions
and innovations have made it possible to create instruction-following,
conversational bots that can perform tasks on behalf of the user. The
thought of using natural language to transform and analyze data is appealing.

## Approaches

When discussed internally at Voltron Data, we identified three primary
approaches to applying LLMs to data that can be implemented today:

1. LLM writes SQL
2. LLM writes a UDF (Python or another language)

> 3. Use LLM in a UDF (that a LLM could write)

Reviewer 2: I think it's better to omit the parenthetical here. Instead add a sentence below saying that these aren't necessarily mutually exclusive, for example the LLM could write the LLM inference code in the UDF definition.
```suggestion
3. Use LLM in a UDF
```

> 

Reviewer 2: Can you add another sentence here to clarify that these three approaches are not necessarily the only approaches? There are surely many others, such as having the LLM generate data.

Reviewer 2: Another suggested addition here: Say that all three of these approaches can be demonstrated using Marvin, and explain briefly what Marvin is.

Author: @ianmcook added a blurb if you want to review that

    # num_e: int = Field(..., description="The number of 'e' vowels.") # <3>
    # num_i: int = Field(..., description="The number of 'i' vowels.") # <3>
    # num_o: int = Field(..., description="The number of 'o' vowels.") # <3>
    # num_u: int = Field(..., description="The number of 'u' vowels.") # <3>
    # num_y: int = Field(..., description="The number of 'y' vowels.") # <3>
    num_total: int = Field(..., description="The total number of vowels.")  # <3>

VowelCounter("hello world")  # <4>
```

1. Additional imports for Pydantic.
2. Configure Marvin to use a cheaper model.
3. A non-deterministic, LLM-powered AI model.
4. Call the AI model on some text.

Then we'll have the LLM write the UDF, just to be fancy:

```{python}
udf = create_udf_from_text(
    "a function named count_vowels_ai that given an input string, calls VowelCounter on it and returns the num_total attribute of that result"
)
print(udf)
exec(udf)
```

```{python}
t5 = t3.mutate(
    species_vowel_count=count_vowels_ai(t3.species),
    island_vowel_count=count_vowels_ai(t3.island),
)
t5
```

Again, notice that in this UDF, unlike the previous example, a LLM is being

> called (possibly several times) per row in the table. This is a very expensive

Reviewer 3: ```suggestion
called (possibly several times) for each row in the table. This is a very expensive
```

```{python}
from rich import print

with open("index.qmd", "r") as f:
    self_text = f.read()

@marvin.ai_model
class Summary(BaseModel):
    """Summary of text."""

    summary_line: str = Field(..., description="The one-line summary of the text.")
    summary_paragraph: str = Field(
        ..., description="The one-paragraph summary of the text."
    )
    conclusion: str = Field(
        ..., description="The conclusion the reader should draw from the text."
    )
    key_points: list[str] = Field(..., description="The key points of the text.")
    critiques: list[str] = Field(
        ..., description="Professional, fair critiques of the text."
    )
    suggested_improvements: list[str] = Field(
        ..., description="Suggested improvements for the text."
    )
    sentiment: float = Field(..., description="The sentiment of the text.")
    sentiment_label: str = Field(..., description="The sentiment label of the text.")
    author_bias: str = Field(..., description="The author bias of the text.")

print(Summary(self_text))
```

## Next steps

You can get involved with [Ibis
Birdbrain](https://github.com/ibis-project/ibis-birdbrain), our open-source data

> & AI project for building next-generating natural language interfaces to data.

Reviewer 3: ```suggestion
& AI project for building next-generation natural language interfaces to data.
```

---
title: LLMs and data
author: "Cody Peterson"
date: "2023-09-27"
freeze: auto
categories:
    - blog
    - ai
    - llm
---

## Introduction

While large-language models (LLMs) have been around for years, recent versions
and innovations have made it possible to create instruction-following,
conversational bots that can perform tasks on behalf of the user. The
thought of using natural language to transform and analyze data is appealing.

## Approaches

When discussed internally at Voltron Data, we identified three distinct
approaches to applying LLMs to data analytics that can be implemented today:

> 1. LLM writes SQL

Reviewer 2: ```suggestion
1. LLM writes an analytic routine (SQL query or dataframe code)
```

Reviewer: Probably don't need the parenthetical. How about just `LLM writes a SQL query or dataframe code`?

Author: yep lemme rework per this suggestion

2. LLM writes a subroutine (Python UDF or another language)
3. Use LLM in a subroutine

While these three approaches are not an exhaustive list of how LLMs can be
applied to data, they can be easily understood and implemented with
Ibis and [Marvin](https://www.askmarvin.ai/welcome/what_is_marvin/), an AI
engineering framework for building natural language interfaces. Together with
these two open-source tools, we can build a natural language interface for
data analytics that supports 18+ backends. But first, let's demonstrate the
three approaches in code.

> ### Approach 1: LLM writes SQL

Reviewer 2: ```suggestion
### Approach 1: LLM writes an analytic routine
```

State of the art (SoTA) LLMs are decent at generating SQL out of the box. We can
be clever to handle errors, retries, and more, but in its simplest form:

```{python}
# | code-fold: true
import ibis  # <1>
import marvin  # <1>

from dotenv import load_dotenv  # <1>

load_dotenv()  # <2>

con = ibis.connect("duckdb://penguins.ddb")  # <3>
t = ibis.examples.penguins.fetch()  # <3>
t = con.create_table("penguins", t.to_pyarrow(), overwrite=True)  # <3>
```

1. Import the libraries we need.
2. Load the environment variable to setup Marvin to call our OpenAI account.
3. Setup the demo datain an Ibis backend.

```{python}
import ibis  # <1>
import marvin  # <1>

from ibis.expr.schema import Schema  # <1>
from ibis.expr.types.relations import Table  # <1>

ibis.options.interactive = True  # <2>
marvin.settings.llm_model = "openai/gpt-4"  # <2>

con = ibis.connect("duckdb://penguins.ddb")  # <3>
t = con.table("penguins")  # <3>
```

> 1. Import Ibis and Marvin.

Reviewer 2: What role is Ibis actually playing here, besides submitting the query and pretty-printing the results? Couldn't this just use DBAPI and achieve the same thing? If not can you explain why not?

Author: - easily handoff to LLMs (notice using Ibis to get the schema, table name, nice error messages to feed back, etc)
- translation across SQL dialects (you could just do this with SQLglot)
- access to Ibis for mixed Python/natural language analysis (use language OR Ibis code, mix and match)
- 18+ data platforms, thus this can be a universal standard for AI + data 
- once working reliably, write Ibis code that (somewhat) avoids the SQL dialect issues and works on 18+ backends
- then way down the line, Substrait via Ibis to really avoid those issues

will add some more explanation here

2. Configure Ibis (interactive) and Marvin (GPT-4).
3. Connect to the data and load a table into a variable.

```{python}
@marvin.ai_fn  # <1>
def _generate_sql_select(
    text: str, table_name: str, table_schema: Schema
) -> str:  # <1>
    """Generate SQL SELECT from text."""  # <1>

def sql_from_text(text: str, t: Table) -> Table:  # <2>
    """Run SQL from text."""  # <2>
    return t.sql(_generate_sql_select(text, t.get_name(), t.schema()).strip(";"))  # <2>
```

1. A non-deterministic, LLM-powered AI function.
2. A deterministic, human-authored function that calls the AI function.

```{python}
t2 = sql_from_text("the unique combination of species and islands", t)
t2
```

```{python}
t3 = sql_from_text(
    "the unique combination of species and islands, with their counts, ordered from highest to lowest, and name that column just 'count'",
    t,
)
t3
```

This works well-enough for simple cases and can be expanded to handle complex
ones. In many scenarios, it may be easier to express a query in English or
another language than to write it in SQL, especially if working across multiple
SQL dialects.

> 

Reviewer 2: What about Ibis? Can we show Birdbrain writing Ibis code here? Or can we briefly explain that it _could_ also write Ibis code or code for other dataframe libraries?

### Approach 2: LLM writes a subroutine

If more complex logic needs to be expressed, SoTA LLMs are also decent at
writing Python and a number of other programming languages. Many data platforms
support user-defined functions (UDFs) in Python or some other language. We'll
stick to scalar Python UDFs via DuckDB to demonstrate the concept:

```{python}
@marvin.ai_fn  # <1>
def _generate_python_function(text: str) -> str:  # <1>
    """Generate a simple, typed, correct Python function from text."""  # <1>

def create_udf_from_text(text: str) -> str:  # <2>
    """Create a UDF from text."""  # <2>
    return f"""
import ibis

@ibis.udf.scalar.python
{_generate_python_function(text)}
""".strip()  # <2>
```

1. A non-deterministic, LLM-powered AI function.
2. A deterministic, human-authored function that calls the AI function.

```{python}
udf = create_udf_from_text(
    "a function named count_vowels that given an input string, returns an int w/ the number of vowels (y_included as a boolean option defaulted to False)"
)
print(udf)

> exec(udf)

Reviewer 4: put the exec inside the `create_udf_from_text` function. Also for due diligence I would mention how this is a total security nightmare if you are passing in untrusted user input.

Author: why move it inside the function? with it outside, you (directly or via the bot this will be called from) can inspect the function before you execute/register it

fyi this post has moved over here: https://ibis-project.github.io/ibis-birdbrain/posts/llms-and-data-pt1/

I'm going to have a series of posts to start off that new project, and will make one here in Ibis to announce. don't want to overwhelm the Ibis blog or start making it all about AI/LLMs

Reviewer 4: Eh, just a bit more encapsulated, but yes that is a nice benefit that is probably worth it. Sure keep it outside. Sounds good

    self_text = f.read()

@marvin.ai_model
class Summary(BaseModel):
    """Summary of text."""

    summary_line: str = Field(..., description="The one-line summary of the text.")
    summary_paragraph: str = Field(
        ..., description="The one-paragraph summary of the text."
    )
    conclusion: str = Field(
        ..., description="The conclusion the reader should draw from the text."
    )
    key_points: list[str] = Field(..., description="The key points of the text.")
    critiques: list[str] = Field(
        ..., description="Professional, fair critiques of the text."
    )
    suggested_improvements: list[str] = Field(
        ..., description="Suggested improvements for the text."
    )
    sentiment: float = Field(..., description="The sentiment of the text.")
    sentiment_label: str = Field(..., description="The sentiment label of the text.")
    author_bias: str = Field(..., description="The author bias of the text.")

print(Summary(self_text))
```

## Next steps

You can get involved with [Ibis
Birdbrain](https://github.com/ibis-project/ibis-birdbrain), our open-source data
& AI project for building next-generation natural language interfaces to data.

> ## Discussions

Reviewer: Is this empty `## Discussions` section intentional?

Author: yeah I'm going to remove it though -- it's before the Giscus discussions at the bottom, but would look weird via RSS and it's obvious w/o a header
